src/redux/books/books.js

```diff
@@ -1,26 +1,106 @@
+const ADD_BOOK_STARTED = 'bookStore/books/ADD_BOOK_STARTED';
+const ADD_BOOK_FAILURE = 'bookStore/books/ADD_BOOK_FAILURE';
+const FETCH_BOOKS_SUCCESS = 'bookStore/books/FETCH_BOOKS_SUCCESS';
+const FETCH_BOOKS_STARTED = 'bookStore/books/FETCH_BOOKS_STARTED';
+const FETCH_BOOKS_FAILURE = 'bookStore/books/FETCH_BOOKS_FAILURE';
 
-export const addBook = (payload) => ({
-  type: ADD_BOOK,
+export const removeBook = (payload) => ({
+  type: REMOVE_BOOK,
   payload,
 });
 
-export const removeBook = (payload) => ({
-  type: REMOVE_BOOK,
+export const fetchBooksSuccess = (payload) => ({
+  type: FETCH_BOOKS_SUCCESS,
+  payload,
+});
+
+export const fetchBookStarted = () => ({
+  type: FETCH_BOOKS_STARTED,
+});
+
+export const fetchBookFailure = (error) => ({
+  type: FETCH_BOOKS_FAILURE,
+  payload: error,
+});
+
+export const addBookSuccess = (payload) => ({
+  type: ADD_BOOK_SUCCESS,
   payload,
 });
 
+export const addBookStarted = () => ({
+  type: ADD_BOOK_STARTED,
+});
+
+export const addBookFailure = (error) => ({
+  type: ADD_BOOK_FAILURE,
+  payload: {
+    error,
+  },
+});
+
+export const fetchBooks = () => (dispatch) => {
+  dispatch(fetchBookStarted());
+  axios.get('https://us-central1-bookstore-api-e63c8.cloudfunctions.net/bookstoreApi/apps/ZFiVXR8MDsWp5znsZ8Qa/books').then((res) => {
+    const books = res.data;
+    const bookData = [];
+    _.forEach(books, ([book], key) => bookData.push({ id: key, ...book }));
+    dispatch(fetchBooksSuccess(bookData));
+  }).catch((err) => {
+    const erroMesage = err.message;
+    dispatch(fetchBookFailure(erroMesage));
+  });
+};
+
+export const addBook = ({ itemId = uuidv4(), title, category }) => (dispatch) => {
+  dispatch(addBookStarted());
+  axios.post('https://us-central1-bookstore-api-e63c8.cloudfunctions.net/bookstoreApi/apps/ZFiVXR8MDsWp5znsZ8Qa/books', {
+    item_id: itemId,
+    title,
+    category,
+  })
+    .then((res) => {
+      dispatch(addBookSuccess(res.data));
+    })
+    .catch((err) => {
+      dispatch(addBookFailure(err.message));
+    });
+};
+
+export const deleteBook = (id) => (dispatch) => {
+  axios.delete(`https://us-central1-bookstore-api-e63c8.cloudfunctions.net/bookstoreApi/apps/ZFiVXR8MDsWp5znsZ8Qa/books/${id}`);
+  dispatch(removeBook(id));
+};
+
 const bookReducer = (state = initialState, { type, payload }) => {
   switch (type) {
+    case ADD_BOOK_STARTED:
+      return state;
+    case FETCH_BOOKS_STARTED:
+      return state;
+    case FETCH_BOOKS_SUCCESS:
+      return payload;
+    case FETCH_BOOKS_FAILURE:
+      return state;
+    case ADD_BOOK_SUCCESS:
+      return payload;
+    case ADD_BOOK_FAILURE:
+      return state;
     case ADD_BOOK:
-      return { books: [...state.books, payload] };
+      return payload;
     case REMOVE_BOOK:
-      return state.books.filter((book) => book.id !== payload.id);
+      const newState = _.filter(state, (book) => book.id !== payload);
+      return newState;
     default:
       return state;
   }
```

Comment:
- [ ] The functionality for adding a book produces an error in your app.

1. `addBook` send the new book to the API ✔️ 
2. Line 72-73 The `rest.data` is returning a string that says-> `Created`
3. Line 73 You are dispaching that string by `addBookSuccess function` and it sends the string to your reducer overwriting your state with that string. ❌ 
4. So now, your `state = Created` and you app crash because your BookList component is expecting for a `state = [...data]`. but it is receiving a string ❌ 

Please, be careful with the cases in your reducer. 🔨